kpops/utils/gen_schema.py

```diff
@@ -0,0 +1,139 @@
+import logging
+from enum import Enum
+from typing import (
+    Annotated,
+    Any,
+    Literal,
+    Optional,
+    Sequence,
+    Union,
+    get_args,
+    get_origin,
+)
+
+from pydantic import Field, schema, schema_json_of
+from pydantic.fields import ModelField
+from pydantic.schema import SkipField
+
+from kpops.cli.pipeline_config import PipelineConfig
+from kpops.cli.registry import _find_classes
+from kpops.components.base_components.pipeline_component import PipelineComponent
+
+
+class SchemaScope(str, Enum):
+    PIPELINE = "pipeline"
+    CONFIG = "config"
+
+
+original_field_schema = schema.field_schema
+
+
+# adapted from https://github.com/tiangolo/fastapi/issues/1378#issuecomment-764966955
+def field_schema(field: ModelField, **kwargs: Any) -> Any:
+    if field.field_info.extra.get("hidden_from_schema"):
+        raise SkipField(f"{field.name} field is being hidden")
+    else:
+        return original_field_schema(field, **kwargs)
+
+
+schema.field_schema = field_schema
+
+log = logging.getLogger("")
+
+
+def is_valid_component(
+    used_fields: set[str], component: type[PipelineComponent]
+) -> bool:
+    """
+    Check whether a PipelineComponent subclass has a valid definition for the schema generation.
+
+    :param used_fields: types defined so far
+    :type used_fields: list[str]
+    :param component: component type to be validated
+    :type component: type[PipelineComponent]
+    :return: Whether component is valid for schema generation
+    :rtype: bool
+    """
+    component_name: str = component.__name__
+    component_type: str = component.get_component_type()
+    schema_type: Optional[ModelField] = component.__fields__.get("schema_type")
```

Comment:
is there a reason why you're using explicit type annotations here? it looks to me that those are unnecessary as they can be easily inferred by the type checker

Reply:
It helps me a bit to read what I've written and to catch/avoid mistakes.

Should I remove them?

Reply:
are you using a type checker in your editor? something like Pylance/Pyright or MyPy? these should both be able to infer the types automatically

Reply:
I would prefer to not have explicit type annotations here unless they really make sense, as it makes it a bit verbose

Reply:
I use Pylance.

Understandable, I've removed them.

Reply:
make sure to configure [Inlay Type Hints](https://devblogs.microsoft.com/python/python-in-visual-studio-code-july-2022-release/#inlay-type-hints) for Pylance